contracts/gol.cairo

```diff
@@ -0,0 +1,450 @@
+@external
+func evolve{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game_id : felt
+    ):
+    alloc_locals
+    # evolve by one generation
+    const generations = 1
+
+    let (caller) = ensure_user()
+    evolve_and_save(
+        game_id=game_id,
+        user=caller,
+        generations=generations
+    ) 
+    reward_user(user=caller)
+
+    return ()
+end
+
+
+# Give life to a specific cell (infinite game).
+@external
+func give_life_to_cell{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        cell_index : felt
+    ):
+    # This does not trigger an evolution. Multiple give_life
+    # operations may be called, building up a shape before
+    # a turn triggers evolution.
+    alloc_locals
+    let (caller) = ensure_user()
+    let (generation, current_game_state) = verify_and_get_for_cell_activation(
+        cell_index=cell_index,
+        user=caller
+    )
+    pay(user=caller, credit_requirement=GIVE_LIFE_CREDIT_REQUIREMENT)
+    activate_cell(
+        generation=generation,
+        caller=caller,
+        cell_index=cell_index,
+        current_state=current_game_state
+    )
+
+    return ()
+end
+
+##########
+
+# Returns a felt of encoded cells for the specified game's generation.
+@view
+func view_game{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game_id : felt,
+        generation : felt
+    ) -> (
+        generation_owner : felt,
+        game_state : felt
+    ):
+
+    let (game_info) = stored_game.read(game_id, generation)
+    return (game_info[0], game_info[1])
+end
+
+# Returns the current generation of a given game.
+@view
+func get_current_generation{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game_id : felt
+    ) -> (
+        generation : felt
+    ):
+    let (generation) = current_generation.read(game_id)
+    return (generation)
+end
+
+##########
+
+func pay{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        user : felt, 
+        credit_requirement : felt
+    ):
+    ERC20._burn(user, Uint256(credit_requirement, 0))
+    return ()
+end
+
+func reward_user{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        user : felt
+    ):
+    ERC20._mint(user, Uint256(1, 0))
+    return ()
+end
+
+func ensure_user{
+        syscall_ptr : felt*
+    }(
+    ) -> (
+        caller : felt
+    ):
+    let (caller) = get_caller_address()
+    with_attr error_message(
+        "User not authenticated"
+    ):
+        assert_not_zero(caller)
+    end
+    return (caller)
+end
+
+func verify_new_game{
```

Comment:
This is an assertion, it's nice to start their names with assert, like`assert_valid_new_game`